csaf-validation/src/main/kotlin/io/github/csaf/sbom/validation/tests/Tests.kt

```diff
@@ -1033,6 +1143,145 @@ object Test6128Translation : Test {
+                } else {
+                    versions.map { Pair(it, it.toIntOrNull()) }.filter { it.second == null }
+                }
+                .map { it.first }
+
+        return if (invalids.isEmpty()) {
+            ValidationSuccessful
+        } else {
+            ValidationFailed(
+                listOf(
+                    "The following versions are invalid because of a mix of integer and semantic versioning: ${invalids.joinToString(", ")}"
+                )
+            )
+        }
+    }
+}
+
+val keywords = listOf("after", "all", "before", "earlier", "later", "prior", "versions")
+val operatorsRegex = """(?)(<|<=|>>=|>)""".toRegex()
+
+/**
+ * Implementation of
+ * [Test 6.1.31](https://docs.oasis-open.org/csaf/csaf/v2.0/os/csaf-v2.0-os.html#6131-version-range-in-product-version).
+ */
+object Test6131VersionRangeInProductVersion : Test {
+    override fun test(doc: Csaf): ValidationResult {
+        var versions =
+            doc.product_tree.mapBranchesNotNull(
+                predicate = { it.category == Csaf.Category3.product_version }
+            ) {
+                it.name
+            }
+        val invalids =
+            versions.filter {
+                operatorsRegex.containsMatchIn(it) ||
+                    keywords.any { kw -> it.split("""\s""".toRegex()).contains(kw) }
+            }
+        return if (invalids.isEmpty()) {
+            ValidationSuccessful
+        } else {
+            ValidationFailed(
+                listOf(
+                    "The following product versions are invalid and contain version ranges: ${invalids.joinToString(", ")}"
+                )
+            )
+        }
+    }
+}
+
+/**
+ * Implementation of
+ * [Test 6.1.31](https://docs.oasis-open.org/csaf/csaf/v2.0/os/csaf-v2.0-os.html#6132-flag-without-product-reference).
+ */
+object Test6132FlatWithoutProductReference : Test {
+    override fun test(doc: Csaf): ValidationResult {
+        val missing =
+            doc.vulnerabilities
+                ?.flatMap { it.flags ?: setOf() }
+                ?.filter { it.group_ids == null && it.product_ids == null } ?: listOf()
+        return if (missing.isEmpty()) {
+            ValidationSuccessful
+        } else {
+            ValidationFailed(
+                listOf(
+                    "The following flags are missing products or groups: ${missing.map { it.label }.joinToString(", ")}"
+                )
+            )
+        }
+    }
+}
+
+/**
+ * Implementation of
+ * [Test 6.1.33](https://docs.oasis-open.org/csaf/csaf/v2.0/os/csaf-v2.0-os.html#6133-multiple-flags-with-vex-justification-codes-per-product).
+ */
+object Test6133MultipleFlagsWithVEXJustificationCodesPerProduct : Test {
+    override fun test(doc: Csaf): ValidationResult {
+        var duplicates = mutableListOf<String>()
+        var groupMap = doc.gatherProductIdsPerGroup()
+
+        for (vuln in doc.vulnerabilities ?: listOf()) {
+            val productsIDsInFlags =
```

Comment:
Shouldn't we add a filter for VEX justification codes in the flag's `label`? (However, no other label is allowed for now)

Reply:
I skipped it because it seems that all labels are VEX justification codes but I can do it just to make it future safe.

Reply:
I already thought so. The standard also states that this is only useful for future releases which may add more labels. If we skip the filter, I'd at least expect a comment/documentation of this decision.